torchtext/datasets/iwslt2016.py

```diff
@@ -225,50 +204,67 @@ def IWSLT2016(root='.data', split=('train', 'valid', 'test'), language_pair=('de
+
     languages = "-".join([src_language, tgt_language])
 
-    iwslt_tar = '{}/{}/texts/{}/{}/{}.tgz'
-    iwslt_tar = iwslt_tar.format(
-        root, SUPPORTED_DATASETS['_PATH'].split(".")[0], src_language, tgt_language, languages)
-    extracted_dataset_tar = extract_archive(iwslt_tar)
-    extracted_files.extend(extracted_dataset_tar)
-
-    # Clean the xml and tag file in the archives
-    file_archives = []
-    for fname in extracted_files:
-        if 'xml' in fname:
-            _clean_xml_file(fname)
-            file_archives.append(os.path.splitext(fname)[0])
-        elif "tags" in fname:
-            _clean_tags_file(fname)
-            file_archives.append(fname.replace('.tags', ''))
-        else:
-            file_archives.append(fname)
-
-    data_filenames = {
-        "train": _construct_filepaths(file_archives, src_train, tgt_train),
-        "valid": _construct_filepaths(file_archives, src_eval, tgt_eval),
-        "test": _construct_filepaths(file_archives, src_test, tgt_test)
+    iwslt_tar = os.path.join(
+        "texts", src_language, tgt_language, languages
+    ) + ".tgz"
+
+    cache_decompressed_dp = cache_compressed_dp.on_disk_cache(
+        filepath_fn=lambda x: os.path.join(root, os.path.splitext(_PATH)[0], iwslt_tar)
+    )
+    cache_decompressed_dp = FileOpener(cache_decompressed_dp, mode="b").read_from_tar().filter(lambda x: iwslt_tar in x[0])
+    cache_decompressed_dp = cache_decompressed_dp.end_caching(mode="wb", same_filepath_fn=True)
+
+    file_path_by_lang_and_split = {
+        src_language: {
+            "train": src_train,
+            "valid": src_eval,
+            "test": src_test,
+        },
+        tgt_language: {
+            "train": tgt_train,
+            "valid": tgt_eval,
+            "test": tgt_test,
+        }
     }
 
-    for key in data_filenames.keys():
-        if len(data_filenames[key]) == 0 or data_filenames[key] is None:
-            raise FileNotFoundError(
-                "Files are not found for data type {}".format(key))
+    src_filepath = file_path_by_lang_and_split[src_language][split]
+
+    cache_inner_src_decompressed_dp = cache_decompressed_dp.on_disk_cache(
+        filepath_fn=lambda x: os.path.join(root, "2016-01/texts/", src_language, tgt_language, languages, src_filepath)
+    )
+    cache_inner_src_decompressed_dp = FileOpener(cache_inner_src_decompressed_dp, mode="b").read_from_tar()
+    cache_inner_src_decompressed_dp = cache_inner_src_decompressed_dp.map(lambda x: _clean_files(x[0], os.path.splitext(os.path.dirname(os.path.dirname(x[0])))[0], x[1]))
+    cache_inner_src_decompressed_dp = cache_inner_src_decompressed_dp.filter(lambda x: src_filepath in x)
+    cache_inner_src_decompressed_dp = FileOpener(cache_inner_src_decompressed_dp, mode="b")
+    cache_inner_src_decompressed_dp = cache_inner_src_decompressed_dp.end_caching(mode="wb", same_filepath_fn=True)
```

Comment:
OK , I think we might need few changes here to optimize the whole thing. 
1) Cache file looks good (we are looking at the end state file after clean-up). But let's use variable  so that we can reuse it later in filtering part. 
2) We should reverse the order of filter and map so that we only clean the file we are interested in. 
3) The _clean_files function should return the stream (instead of saving it to file) so that that can be used to save the actual clean file through end_caching datapipe.

Also this mechanism is same for both src and trg, let's create a inner function to reuse across both the datapipes.

Reply:
One challenge with #2 is that the filenames get changed in the `_clean_inner*` functions, so we don't "know" them a priori. I haven't given much thought yet to whether we *could* know them...

Reply:
OK, I have applied similar logic to the "expected files" to create a mapping for "uncleaned files" and this seems like a nice solution. I am still returning the output file (i.e., the `filepath_fn` file) from `_clean_files` and it seems OK, but there doesn't seem to be any caching (subsequent calls to `next(iter(train))` take the same time).

Reply:
Scratch that -- I fixed it. 🚀  First read is ~2.6s and subsequent reads are ~600ms. It still seems like a lot of time, but maybe there are other things I've missed... This is starting to look a lot better now.